mathjax3-ts/adaptors/HTMLAdaptor.ts

```diff
@@ -0,0 +1,430 @@
+    document: D;
+    DOMParser: {
+        new(): MinDOMParser<D>
+    };
+    NodeList: any;
+    HTMLCollection: any;
+    HTMLElement: any;
+    DocumentFragment: any;
+    Document: any;
+}
+
+/*****************************************************************/
+/*
+ * The minimum needed for an HTML Adaptor
+ */
+export interface MinHTMLAdaptor<N, T, D> extends DOMAdaptor<N, T, D> {
+    window: MinWindow<D>
+}
+
+/*****************************************************************/
+/*
+ *  Abstract HTMLAdaptor class for manipulating HTML elements
+ *  (subclass of AbstractDOMAdaptor)
+ *
+ *  N = HTMLElement node class
+ *  T = Text node class
+ *  D = Document class
+ */
+
+export class HTMLAdaptor<N extends MinHTMLElement<N, T>,
+                         T extends MinText<N, T>,
+                         D extends MinDocument<N, T>>
+extends AbstractDOMAdaptor<N, T, D> implements MinHTMLAdaptor<N, T, D> {
+    /*
+     * The window object for this adaptor
+     */
+    window: MinWindow<D>;
+
+    /*
+     * The DOMParser used to parse a string into a DOM tree
+     */
+    parser: MinDOMParser<D>;
+
+    /*
+     * @override
+     * @constructor
+     */
+    constructor(window: MinWindow<D>) {
+        super(window.document);
+        this.window = window;
+        this.parser = new (window.DOMParser as any)();
+    }
+
+    /*
+     * @override
+     */
+    public parse(text: string, format: string = 'text/html') {
+        return this.parser.parseFromString(text, format);
+    }
+
+    /*
+     * @override
+     */
+    protected create(type: string) {
+        return this.document.createElement(type);
+    }
+
+    /*
+     * @override
+     */
+    public text(text: string) {
+        return this.document.createTextNode(text);
+    }
+
+    /*
+     * @override
+     */
+    public head(doc: D) {
+        return doc.head;
+    }
+
+    /*
+     * @override
+     */
+    public body(doc: D) {
+        return doc.body;
+    }
+
+    /*
+     * @override
+     */
+    public root(doc: D) {
+        return doc.documentElement;
+    }
+
+    /*
+     * @override
+     */
+    public tags(node: N, name: string, ns: string = null) {
+        let nodes = (ns ? node.getElementsByTagNameNS(ns, name) : node.getElementsByTagName(name));
+        return Array.from(nodes as N[]) as N[];
+    }
+
+    /*
+     * @override
+     */
+    public getElements(nodes: (string | N | N[])[], document: D) {
+        let containers: N[] = [];
+        for (const node of nodes) {
+            if (typeof(node) === 'string') {
+                containers = containers.concat(Array.from(this.document.querySelectorAll(node)));
+            } else if (Array.isArray(node)) {
+                containers = containers.concat(Array.from(node) as N[]);
+            } else if (node instanceof this.window.NodeList || node instanceof this.window.HTMLCollection) {
+                containers = containers.concat(Array.from(node as any as N[]));
```

Comment:
This is because Typescript apparently doesn't realize that NodeList and HTMLCollection are iterables (even though they are), so won't let me call `Array.from()` on them directly.  So I have to case them to something iterable, like an array.  Since I want the array to be `N[]` anyway (both `NodeList` and `HTMLCollection` will be arrays of `N`), I want to cast to that.  But Typescript won't let me cast to `N[]` directly (it doesn't realize the the types are compatible), and so I cast to `any` first, and then to `N[]`.  This is all just to get around the fact that Typescript doesn't properly know the types of the things involved.